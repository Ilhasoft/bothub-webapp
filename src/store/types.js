export default {
  SET_TOKEN: 'SET_TOKEN',
  OPEN_LOGIN_MODAL: 'OPEN_LOGIN_MODAL',
  CLOSE_LOGIN_MODAL: 'CLOSE_LOGIN_MODAL',
  SET_PROFILE: 'SET_PROFILE',
  SET_NICKNAME_AUTHENTICATED: 'SET_NICKNAME_AUTHENTICATED',
  SET_REPOSITORY_RELATED_UUID: 'SET_REPOSITORY_RELATED_UUID',
  CACHE_FETCH: 'CACHE_FETCH',
};
