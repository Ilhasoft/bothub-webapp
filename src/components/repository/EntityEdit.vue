<template>
  <div class="entity-edit__entities-list">
    <div class="entity-edit__title entity-edit__entities-list__header">
      <div class="entity-edit__title entity-edit__entities-list__header__container">
        {{ $t('webapp.home.entities_list') }}
        <div
          class="entity-edit__title entity-edit__entities-list__header__question">
          <b-tooltip
            :label="$t('webapp.summary.entity_question')"
            class="tooltipStyle"
            multilined
            type="is-dark"
            position="is-right">
            <b-icon
              custom-size="mdi-18px"
              type="is-dark"
              icon="help-circle"
            />
          </b-tooltip>
        </div>
      </div>
      <b-loading
        :is-full-page="false"
        :active="loading" />
      <b-field
        class="entity-edit__button-group"
        grouped>
        <b-button
          v-if="canEdit"
          ref="editButton"
          class="entity-edit__button"
          type="is-primary"
          @click.native="editing=!editing">
          {{ editing ? $t('webapp.home.finish_editing') : $t('webapp.home.edit_groups') }}
        </b-button>
        <b-button
          v-show="editing"
          :disabled="creating"
          class="entity-edit__button"
          type="is-secondary"
          @click="creating = true">
          {{ $t('webapp.home.create_new_group') }}
        </b-button>
      </b-field>
    </div>
    <div>
      <div
        class="entity-edit__entities-list__labeled-count">
        <span v-show="groups.length > 0">
          {{ $tc('webapp.home.entities_label', groups.length, ungrouped.length) }}
        </span>
        <span v-show="ungrouped.length > 0">
          <span v-if="groups.length > 0">
            {{ $tc('webapp.home.non-grouped_entities', ungrouped.length) }}
          </span>
          <span v-else>
            {{ $tc('webapp.home.only_non-grouped_entities', ungrouped.length) }}
          </span>
        </span>
      </div>
      <create-badges-card
        v-if="creating"
        :title="$t('webapp.home.new_group_named', {})"
        format
        @close="creating = false"
        @finished="createGroup"
      />
      <badges-card-drag-drop
        v-for="group in groups"
        :identifier="group.group_id"
        :key="`${group.group_id}-${needsUpdate}`"
        :list="group.entities"
        :title="$tc('webapp.home.labeled', group.entities.length, { label_value: group.value })"
        :examples-count="0"
        :edit="editing"
        closable
        @onMove="moveEntity"
        @onRemoveCard="onRemoveGroup(group)"
        @onCloseTag="onRemoveEntity($event, group.group_id)"
      />
    </div>
    <badges-card-drag-drop
      :key="needsUpdate"
      :list="ungrouped"
      :examples-count="0"
      :edit="editing"
      :title="$tc('webapp.home.unlabeled', ungrouped.length)"
      identifier="ungrouped"
      dark
      @onCloseTag="onRemoveEntity($event, null)"
      @onMove="moveEntity"
    />
  </div>

</template>

<script>
import BadgesCardDragDrop from '@/components/repository/BadgesCardDragDrop';
import CreateBadgesCard from '@/components/repository/CreateBadgesCard';
import { mapActions, mapGetters } from 'vuex';

export default {
  name: 'EntityEdit',
  components: {
    BadgesCardDragDrop,
    CreateBadgesCard,
  },
  props: {
    groups: {
      type: Array,
      default: () => [],
    },
    ungrouped: {
      type: Array,
      default: () => [],
    },
    repositoryUuid: {
      type: String,
      required: true,
    },
    canEdit: {
      type: Boolean,
      default: true,
    },
  },
  data() {
    return {
      creating: false,
      editing: false,
      loading: false,
      needsUpdate: false,
    };
  },
  computed: {
    ...mapGetters({
      version: 'getSelectedVersion',
    }),
  },
  watch: {
    editing() {
      this.creating = false;
    },
  },
  methods: {
    ...mapActions([
      'editEntity',
      'addGroup',
      'deleteEntity',
      'deleteGroup',
    ]),
    showError(error) {
      let message = this.$t('webapp.home.default_error');

      if (error.response
      && error.response.data
      && error.response.data.non_field_errors
      && error.response.data.non_field_errors.length > 0) message = error.response.data.non_field_errors.join(', ');

      this.$buefy.toast.open({
        message,
        type: 'is-danger',
      });
    },
    async createGroup(text) {
      if (!text || text.length === 0) return;
      this.loading = true;
      try {
        const newGroup = await this.addGroup({
          name: text,
          repositoryId: this.repositoryUuid,
          version: this.version,
        });
        this.$emit('createdGroup', { ...newGroup.data, entities: [], group_id: newGroup.data.id });
        this.creating = false;
      } catch (e) {
        this.showError(e);
      } finally {
        this.loading = false;
      }
    },
    updateGroups(from, to, targetList, sourceList) {
      if (from !== 'ungrouped') {
        this.$emit('updateGroup', { entities: sourceList, groupId: from });
      } else {
        this.$emit('updateUngrouped', { entities: sourceList });
      }

      if (to !== 'ungrouped') {
        this.$emit('updateGroup', { entities: targetList, groupId: to });
      } else {
        this.$emit('updateUngrouped', { entities: targetList });
      }
    },
    async removeEntity(entity, groupId) {
      this.loading = true;
      try {
        await this.deleteEntity({
          entityId: entity.entity_id,
        });
        this.$emit('removeEntity', { entity, groupId });
      } catch (e) {
        this.showError(e);
      } finally {
        this.loading = false;
      }
    },
    async removeGroup(groupId) {
      this.loading = true;
      try {
        await this.deleteGroup({
          groupUuid: groupId,
        });
        this.$emit('removeGroup', groupId);
      } catch (e) {
        this.showError(e);
      } finally {
        this.loading = false;
      }
    },
    async moveEntity(event) {
      const {
        entity, from, to, targetList, sourceList,
      } = event;
      this.loading = true;
      try {
        await this.editEntity({
          entityId: entity.entity_id,
          name: entity.value,
          repositoryVersion: this.version,
          repositoryId: this.repositoryUuid,
          groupId: to === 'ungrouped' ? null : to,
        });
        this.updateGroups(from, to, targetList, sourceList);
      } catch (e) {
        this.showError(e);
        this.needsUpdate = !this.needsUpdate;
      } finally {
        this.loading = false;
      }
    },
    onRemoveEntity(entity, groupId) {
      this.$buefy.dialog.alert({
        title: this.$t('webapp.home.delete_entity'),
        message: this.$t('webapp.home.delete_entity_message', { entity: entity.value }),
        confirmText: this.$t('webapp.home.delete'),
        cancelText: this.$t('webapp.home.cancel'),
        canCancel: true,
        closeOnConfirm: true,
        type: 'is-danger',
        onConfirm: async () => {
          this.removeEntity(entity, groupId);
        },
      });
    },
    onRemoveGroup(group) {
      this.$buefy.dialog.alert({
        title: this.$t('webapp.home.delete_group'),
        message: this.$t('webapp.home.delete_group_message', { group: group.value }),
        confirmText: this.$t('webapp.home.delete'),
        cancelText: this.$t('webapp.home.cancel'),
        canCancel: true,
        closeOnConfirm: true,
        type: 'is-danger',
        onConfirm: async () => {
          this.removeGroup(group.group_id);
        },
      });
    },
  },
};
</script>

<style lang="scss" scoped>
@import '~@/assets/scss/variables.scss';
@import '~@/assets/scss/colors.scss';

.entity-edit {
  &__title {
    font-size: 1.75rem;
    font-weight: $font-weight-medium;
    font-family: $font-family;
    color: $color-fake-black;
  }

  &__button {
    color: white;
    margin-left: 0.5rem;
    margin-top:0.6rem;
  }

  &__entities-list {
    padding: 1rem .5rem;
    &__header {
      display: flex;
      justify-content: space-between;
      flex-wrap: wrap;
      align-items: center;

      &__container{
      display: flex;
      }
      &__question{
      margin: -0.1rem 0 0 0.2rem;
      }
    }
  }

  &__entities-list {
    &__labeled-count {
      margin-top: $between-title-subtitle;
      margin-bottom: $between-subtitle-content;
    }
  }
}

.tooltipStyle::after {
  font-size: 12px;
  line-height: 13px;
  padding: 1rem 0.5rem;
  font-family: $font-family;
}
</style>
