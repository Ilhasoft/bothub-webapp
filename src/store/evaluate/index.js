import actions from './actions';
import state from './state';
import mutations from './mutations';
import getters from './getters';
import types from './types';

export default {
  actions,
  state,
  mutations,
  getters,
  types,
};
