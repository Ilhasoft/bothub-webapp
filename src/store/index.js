import Vue from 'vue';
import Vuex from 'vuex';

import Auth from './auth';
import User from './user';
import Org from './org';
import Repository from './repository';
import Category from './category';
import Example from './example';
import EvaluateExample from './evaluate-example';
import Translate from './translate';
import CachedFetch from './cached-fetch';
import Entity from './entity';
import Intent from './intent';
import News from './news';
import Tutorial from './tutorial';
import External from './external';

Vue.use(Vuex);

const store = new Vuex.Store({
  modules: {
    Auth,
    User,
    Org,
    Repository,
    Category,
    Example,
    Translate,
    CachedFetch,
    EvaluateExample,
    Entity,
    News,
    Tutorial,
    External,
    Intent,
  },
});

store.dispatch('retriveAuthToken');

export default store;
