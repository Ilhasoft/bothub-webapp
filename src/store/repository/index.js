import state from './state';
import getters from './getters';
import actions from './actions';
import mutations from './mutations';
import types from './types';

export default {
  state,
  getters,
  actions,
  mutations,
  types,
};
