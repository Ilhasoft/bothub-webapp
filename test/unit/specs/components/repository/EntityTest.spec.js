import { shallowMount, createLocalVue } from '@vue/test-utils';
import EntityEdit from '@/components/repository/EntityEdit';
import store from '@/store';

const localVue = createLocalVue();

describe('EntityEdit', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(EntityEdit, {
      localVue,
      store,
      mocks: {
        $t: () => 'some specific text',
        $tc: () => 'some specific text',
      },
      propsData: {
        groups: [
          { group_id: 25, value: 'Animal', entities: [{ entity_id: 33, value: 'cat' }, { entity_id: 36, value: 'badger' }, { entity_id: 43, value: 'possum' }] },
          { group_id: 66, value: 'Flower', entities: [{ entity_id: 64, value: 'amelia' }, { entity_id: 69, value: 'daisy' }, { entity_id: 24, value: 'dandelion' }] },
        ],
        ungrouped: [
          { group_id: 78, value: 'car' }, { entity_id: 54, value: 'disease' }, { entity_id: 55, value: 'my_leg' },
        ],
        canEdit: true,
        repositoryUuid: '123abc',
      },
    });
  });

  test('mount', () => {
    expect(wrapper.vm).toBeDefined();
  });

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('editing button works', () => {
    beforeEach(() => {
      const editButton = wrapper.find({ ref: 'editButton' });
      editButton.trigger('click');
    });
    test('editing mode activates correctly', () => {
      expect(wrapper.find({ ref: 'editButton' })).toBeTruthy();
    });
  });
});
