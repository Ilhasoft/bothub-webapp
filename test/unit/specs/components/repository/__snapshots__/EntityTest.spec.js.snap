// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityEdit renders correctly 1`] = `
<div class="entity-edit__entities-list">
  <div class="entity-edit__title entity-edit__entities-list__header">
    <div class="entity-edit__title entity-edit__entities-list__header__container">
      some specific text
      <div class="entity-edit__title entity-edit__entities-list__header__question">
        <b-tooltip-stub active="true" type="is-dark" label="some specific text" position="is-right" multilined="true" size="is-medium" class="tooltipStyle">
          <b-icon-stub type="is-dark" icon="help-circle" customsize="mdi-18px"></b-icon-stub>
        </b-tooltip-stub>
      </div>
    </div>
    <b-loading-stub animation="fade" oncancel="function _default() {}"></b-loading-stub>
    <b-field-stub grouped="true" addons="true" class="entity-edit__button-group">
      <b-button-stub type="is-primary" nativetype="button" tag="button" class="entity-edit__button">
        some specific text
      </b-button-stub>
      <b-button-stub type="is-secondary" nativetype="button" tag="button" class="entity-edit__button" style="display: none;">
        some specific text
      </b-button-stub>
    </b-field-stub>
  </div>
  <div>
    <div class="entity-edit__entities-list__labeled-count"><span>
        some specific text
      </span> <span><span>
          some specific text
        </span></span></div>
    <!---->
    <badges-card-drag-drop-stub title="some specific text" list="[object Object],[object Object],[object Object]" examplescount="0" closable="true" identifier="25"></badges-card-drag-drop-stub>
    <badges-card-drag-drop-stub title="some specific text" list="[object Object],[object Object],[object Object]" examplescount="0" closable="true" identifier="66"></badges-card-drag-drop-stub>
  </div>
  <badges-card-drag-drop-stub title="some specific text" list="[object Object],[object Object],[object Object]" examplescount="0" dark="true" identifier="ungrouped"></badges-card-drag-drop-stub>
</div>
`;
